pkg/overlay/server.go

```diff
@@ -128,49 +159,151 @@ func (server *Server) FindStorageNodes(ctx context.Context, req *pb.FindStorageN
 }
 
 // TODO: nicer method arguments
-func (server *Server) populate(ctx context.Context,
-	startID storj.NodeID, maxNodes int64,
-	minRestrictions *pb.NodeRestrictions,
-	minReputation *pb.NodeStats,
-	excluded storj.NodeIDList) ([]*pb.Node, storj.NodeID, error) {
-
-	// TODO: move the query into db
-	limit := int(maxNodes * 2)
-	nodes, err := server.cache.db.List(ctx, startID, limit)
+// filterNodes removes nodes with duplicate IP addresses and updates excluded list with all nodes
+func (server *Server) filterNodes(ctx context.Context, nodes []*pb.Node, nodeRequirement int64, usedAddrs map[string]bool,
+	excluded storj.NodeIDList) (resultNodes []*pb.Node, updatedExcluded storj.NodeIDList, updatedUsed map[string]bool) {
+
+	for _, n := range nodes {
+		addr := n.Address.GetAddress()
+
+		excluded = append(excluded, n.Id) // exclude all nodes on next iteration
+		updatedExcluded = excluded
+		if !usedAddrs[addr] {
+			resultNodes = append(resultNodes, n)
+			usedAddrs[addr] = true
+		}
+	}
+	return resultNodes, updatedExcluded, usedAddrs
+}
+
+func (server *Server) getReputableNodes(ctx context.Context, startID storj.NodeID, maxNodes int64,
+	minRestrictions *pb.NodeRestrictions, excluded storj.NodeIDList) ([]*pb.Node, storj.NodeID, error) {
+
+	limit := int(float64(maxNodes) * server.matchingNodeRatio)
+	minReputation := server.minStats
+
+	nodeList, err := server.cache.db.List(ctx, startID, limit)
 	if err != nil {
 		server.log.Error("Error listing nodes", zap.Error(err))
 		return nil, storj.NodeID{}, Error.Wrap(err)
 	}
 
-	var nextStart storj.NodeID
-	result := []*pb.Node{}
+	if len(nodeList) <= 0 {
+		server.log.Info("No Keys returned from List operation")
+		return []*pb.Node{}, startID, nil
+	}
+
+	var nodeIDs []storj.NodeID
+	for _, node := range nodeList {
+		nodeIDs = append(nodeIDs, node.Id)
+	}
+
+	nodes, err := server.cache.db.GetAll(ctx, nodeIDs)
+	if err != nil {
+		server.log.Error("Error getting nodes", zap.Error(err))
+		return nil, storj.NodeID{}, Error.Wrap(err)
+	}
+
+	var resultNodes []*pb.Node
+
 	for _, v := range nodes {
 		if v == nil {
 			continue
 		}
 
-		nextStart = v.Id
 		if v.Type != pb.NodeType_STORAGE {
+			server.log.Debug("not storage node = " + v.Id.String() + " was " + v.Type.String())
 			continue
 		}
 
 		restrictions := v.GetRestrictions()
 		reputation := v.GetReputation()
 
+		if reputation.GetAuditCount() < server.newNodeAuditThreshold {
+			continue
+		}
+
 		if restrictions.GetFreeBandwidth() < minRestrictions.GetFreeBandwidth() ||
```

Comment:
let's create a separate method for checking this, rather than having a large if statement.